src/main/java/com/elearningbackend/service/QuestionService.java

```diff
@@ -0,0 +1,232 @@
+                throw new ElearningException(Errors.QUESTION_BANK_PRARENT_NOT_EXISTS.getId(),
+                        Errors.QUESTION_BANK_PRARENT_NOT_EXISTS.getMessage());
+            }
+        }
+    }
+
+    private void validateQuestionType(QuestionDto questionDto) throws ElearningException {
+        if(questionDto.getQuestionBankDto().getQuestionType() == Constants.Q_TYPE_CHOOSE_ONE
+                || questionDto.getQuestionBankDto().getQuestionType() == Constants.Q_TYPE_CHOOSE_MULTIPLE )
+        {
+            if(questionDto.getAnswerDtos().size()<Constants.LENGTH_ANSWER_DEFAULT)
+                throw new ElearningException(Errors.ANSWER_TOO_LESS.getId(),
+                        Errors.ANSWER_TOO_LESS.getMessage());
+            if(!hasCorrectAnswer(questionDto.getAnswerDtos()))
+                throw new ElearningException(Errors.NOT_ANSWER_CORRECT.getId(),
+                        Errors.NOT_ANSWER_CORRECT.getMessage());
+        }
+    }
+
+    private void formatChildQuestion(List<QuestionDto> questionDtos, QuestionDto questionDto, QuestionBankDto questionBankDto,String oldQuestionCode){
+        /**
+         * Check this question has child
+         * if result is true
+         * => change children's question code by new parent's question code
+         * Purpose: don't set parent's question code for each children question
+         */
+        if(hasChildQuestion(questionDtos,questionBankDto.getQuestionCode())){
+            changeQuestionParentCode(questionDtos,oldQuestionCode
+                    , questionDto.getQuestionBankDto().getQuestionCode());
+        }
+    }
+
+    private boolean hasCorrectAnswer(List<AnswerDto> answerDtos){
+        return answerDtos.stream().anyMatch(e -> e.getSystemResultDto().getSystemResultIsCorrect()==Constants.ANSWER_CORRECT);
+    }
+
+    private boolean hasChildQuestion(List<QuestionDto> questionDtos, String questiongCode){
+        return questionDtos.stream().anyMatch(e -> e.getQuestionBankDto().getQuestionParentCode()!=null && e.getQuestionBankDto().getQuestionParentCode().equals(questiongCode));
+    }
+
+    private void changeQuestionParentCode(List<QuestionDto> questionDtos, String oldQuestionCode, String newQuestionCode){
+        questionDtos.stream().map(e -> {
+            if(e.getQuestionBankDto().getQuestionCode().equals(oldQuestionCode)){
+                e.getQuestionBankDto().setQuestionCode(newQuestionCode);
+            }
+            return e;
+        });
+    }
+
+    private List<AnswerDto> addAnswerDtoAndSystemResult(List<SystemResultDto> systemResultDtos,int isCorrect,int length){
+        List<AnswerDto> answerDtos= new ArrayList<>();
+        AnswerDto answerDto = null;
+        for (int i = 0, size = systemResultDtos.size(); i < size; i++) {
+            if(length > 0 && i >length)
+                break;
+            if(isCorrect==Constants.ANSWER_INCORRECT && systemResultDtos.get(i).getSystemResultIsCorrect() != Constants.ANSWER_INCORRECT)
```

Comment:
Cái chỗ này kinh quá em ơi. Đọc mãi mới luận ra được. Chỗ này có phải check nếu muốn lấy ra toàn bộ answer incorrect và phần tử thứ i của systemResultDtos mà có is_correct khác Constants.ANSWER_INCORRECT thì skip đúng ko? Chỗ này nên tách ra hàm riêng. Ở dưới cũng như vậy, refactor 2 thằng dùng 1 hàm nhé